qcengine/programs/openmm.py

```diff
@@ -0,0 +1,194 @@
+            rw_mol.AddAtom(Chem.Atom(sym.title()))
+
+        # Add in connectivity
+        bond_types = {1: Chem.BondType.SINGLE, 2: Chem.BondType.DOUBLE, 3: Chem.BondType.TRIPLE}
+        for atom1, atom2, bo in jmol.connectivity:
+            rw_mol.AddBond(atom1, atom2, bond_types[bo])
+
+        mol = rw_mol.GetMol()
+
+        # Write out the conformer
+        natom = len(jmol.symbols)
+        conf = Chem.Conformer(natom)
+        bohr2ang = ureg.conversion_factor("bohr", "angstrom")
+        for line in range(natom):
+            conf.SetAtomPosition(line, (bohr2ang * jmol.geometry[line, 0],
+                                        bohr2ang * jmol.geometry[line, 1],
+                                        bohr2ang * jmol.geometry[line, 2]))  # yapf: disable
+
+        mol.AddConformer(conf)
+        Chem.rdmolops.SanitizeMol(mol)
+
+        return mol
+
+    def compute(self, input_data: 'ResultInput', config: 'JobConfig') -> 'Result':
+        """
+        Runs OpenMM on given structure, inputs, in vacuum.
+        """
+        self.found(raise_error=True)
+
+        from simtk import openmm
+        from simtk import unit
+
+        from openforcefield.topology import Molecule
+        from openforcefield.typing.engines.smirnoff import ForceField
+        
+        # Failure flag
+        ret_data = {"success": False}
+
+        # get number of threads to use from `JobConfig.ncores`; otherwise, try environment variable
+        nthreads = config.ncores
+        if nthreads is None:
+            nthreads = os.environ.get('OPENMM_CPU_THREADS')
+
+        # Set workdir to scratch
+        # Location resolution order config.scratch_dir, /tmp
+        parent = config.scratch_directory
+        with temporary_directory(parent=parent, suffix="_openmm_scratch") as tmpdir:
+
+            # Grab molecule, forcefield
+            jmol = input_data.molecule
+            
+            # TODO: If urls are supported by
+            # `openforcefield.typing.engines.smirnoff.ForceField` already, we
+            # can eliminate the `offxml` and `url` distinction
+            # URL processing can happen there
+            if input_data.model.offxml:
+                offxml = input_data.model.offxml
+            elif input_data.model.url:
+                # TODO: add pull of offxml data from a url
+                pass
+            else:
+                raise InputError("OpenMM requires either `model.offxml` or `model.url` to be set")
+
+            # Process molecule with RDKit
+            rdkit_mol = self._process_molecule_rdkit(jmol)
+
+            # Create an OFF Molecule from the RDKit Molecule
+            off_mol = Molecule(rdkit_mol)
+
+            # Load an OFF Force Field
+            off_forcefield = ForceField(offxml)
+
+            # Create OpenMM system in vacuum from forcefield, molecule
+            off_top = off_mol.to_topology()
+            openmm_system = off_forcefield.create_openmm_system(off_top)
```

Comment:
If we are asked to compute many gradients/energies for the same molecule, this step can also introduce a significant amount of overhead if we call it over and over again. Can we also memoize/cache the `openmm_system` object if different geometries for the same molecule are called over and over again, or is that difficult because the bonds and bond orders are being perceived from geometry? 

I imagine that we could use the (optional) SMILES field---if these are specified, it should be easy to memoize/cache the `System` object in terms of this SMILES field.

Reply:
See question above re: where QCEngine implements caching of computed components for later use. I can them try the approach you've laid out here and we can discuss if this meets our needs or introduces issues (e.g. bond orders). Thanks!

@dgasmith, do you have insights on the question of caching `compute` intermediates from `*Harness` instances?

Reply:
You could consider building a small LRU cache where the you can use a `Molecule.get_hash() + offxml_uid` to build a unique identifier.

Reply:
Lets update this key to `{mol_hash}-{input_model.method}-{hash(forcefield_schema)}`. Here we will need to change the `mol_hash` to only the symbols + connectivity fields. Since we are using RDKit here anyways, I assume that there is some sort of uniqueness perception available at the right abstraction level (connectivity only). `Molecule.get_hash()` will return a unique hash whenever the cartesian geometry changes so this wouldn't work for a geometry optimization.

On the `lru_cache` I would recommend a depth of maybe ~10. I don't know how big they are in memory, but unless I am doing a geometry optimization I am unlikely to get multiple hits per instantiation.

Reply:
@dgasmith do you know of a "uniqueness perception available at the right abstraction level (connectivity only)" that makes sense to use here? My knowledge of the tooling available to give the right representation with the guarantees we need is limited here.

Reply:
I think this should just be symbols + connectivity? Something like:
```
>>> import qcelemental as qcel
>>> caffiene = qcel.models.Molecule.from_data("pubchem:benzene")
	Searching PubChem database for benzene (single best match returned)
	Found 1 result(s)
>>> import json
>>> hash(json.dumps({"connectivity": caffiene.connectivity, "symbols": list(caffiene.symbols)}))
-1973773897279072458
```

You may need to add charge/multiplicity here too, the fields are whatever is used to generate the OpenMM system. Also note that this particular molecule doesn't have connectivity so you may need to add that as perceived by OpenMM as well.